Make compatible with new keras Callback model

ali-tny opened the pull request:
see https://keras.io/guides/writing_your_own_callbacks/ for guide - methods have been split to provide a different one for train/eval/predict time.

This change translates `on_batch_end` to `on_train_batch_end` rather than also for validation batches, so batch history only includes train batch loss - not sure if you wanna include the validation batches in the history here for plotting? Would just need to set `on_test_batch_end` to the same as `on_train_batch_end`.

Commit:
Make compatible with new keras Callback model

see https://keras.io/guides/writing_your_own_callbacks/ for guide -
methods have been split to provide a different one for
train/eval/predict time.

This change translates `on_batch_end` to `on_train_batch_end` rather
than also for validation batches, so batch history only includes train
batch loss.

ig248 reviewed (changes requested):
Could you indicate what versions of keras the old vs new api would be compatible to? Also, maybe update the `Callback` docstring - tbh the only reason the base class is here is to avoid importing it from `keras`, hence avoiding `keras` as a dependency. However, if there are API differences, we could specify a keras version range in the setup.py?

Commit:
Remove typeshed dev dependency and bump versions

typeshed was recently removed from PyPI (and comes bundled with mypy
anyway) so was causing dependency resolution to fail.

ali-tny commented:
I've updated to be backwards compatible (much [like tensorflow.keras does](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/callbacks.py)). I considered just adding the keras dependency and importing the Callback base class, but to install tf.keras it needs all of tensorflow, and it seems kinda heavyweight now that its compatible across versions.

I also added a test - it requires adding matplotlib & ipython as dev dependencies though so it can import `kerasplothistory.callbacks`

Commit:
Add test for compatibility with both APIs
